### ISSUE TITLE
PDO string conversion of floating values not reliable

### ISSUE THREAD

**Issue by xfra35** [bcosca/fatfree-core#106]
Since PDO doesn't handle floating points, the framework has to [cast](https://github.com/bcosca/fatfree-core/blob/c9681ed5b2b5ec34abca1afd9be275d469aa24ec/db/sql.php#L116) float values to string.

This work well with locales having `.` as a decimal separator, but not for those having `,`:
```php
$f3->LANGUAGE='fr';
$price=12.34;
$mapper->price=$price;
var_dump($mapper->price);// '12,34' <= there shouldn't be a comma here
$mapper->save();// <= value stored in database = 12
```
The only way to have this working is to perform a custom string cast before each call to `set` or to overload the `set()` method:
```php
$mapper->price=str_replace(',','.',$price);
var_dump($mapper->price);// '12.34' <= OK
```
This is not convenient and can lead to mistakes.

I think that the framework should handle it, but I'm not sure how. Maybe by adding a `float` flag in the `$db->schema` result?

Any idea?

**Comment by geniuswebtools** [comment]
I think the solution for this would be in the beforeSave() trigger.  The reason I say that is because saving the result of 12 is most likely the result of the database engine and the data type associated with that column.  

I'll assume you're not using SQLite because I'd wager that the SQLite would save the 12,34 value as passed because it doesn't truncated values like the MySQL decimal type does.

**Comment by ikkez** [comment]
So is this a number format/cast issue related to the system locale? I think the mapper only handles floats as string because the topic can be quite complex and full of traps:
* float/integer limitation on 32bit systems
* there is not native `PDO::PARAM_FLOAT`
* possibly rounding issues
* casting locale numbers with thousand separators and decimals can be tricky:
  * DE: `1.234,99`
  * EN: `1,234.99`

But if you set a float value to a mapper and save it to a float or decimal db table column, it should not be trimmed. Is that due to your `fr` language setting?

**Comment by xfra35** [comment]
@geniuswebtools is right, the value gets truncated because MySQL expects a dot as a separator, not a comma. I haven't tried yet with other SQL engines.

> Is that due to your fr language setting?

Yes, and I expect the same behaviour with `de`. Haven't you face the problem already?

**Comment by ikkez** [comment]
Tested. You are right. When the float value is casted to a string, it depends on the current system locale how it's going to be formatted. I suggest to use a custom reliable format here:

```php
$mapper->price = number_format(2234.99, 2, '.', '');
```

**Comment by MINORITYmaN** [comment]
h & happy new year!
i would mention ikkez suggestion in the docs somehow to avoid headaches and tickets;)

cheers

**Comment by xfra35** [comment]
`number_format` is not necessary as the cast to string doesn't introduce any thousand separator. Also it requires to know in advance the number of significant decimals (2 is fine for prices, but not for other floating values). That's why so far I've always processed float values with `str_replace`.

Anyway the question is: how could the framework ease that part and make the storage of floating values a breeze? The statement `$mapper->price=12.34` should be straightforward.

**Comment by ikkez** [comment]
Well then I think `str_replace` is fine. But if one is setting a formatted string as float, it can become problematic. Maybe a preg_replace (find all digits and the last seperator) (like [this one](http://codepad.org/NW4e9hQH) but simplified) could aid here in the mapper setter. Nevertheless, the value should be set as string (all stackoverflow threads I've found recommended that) internally to the PDO query. So when you set `$mapper->price=12.34` we should know when to convert it to a clean float and back to string. This can only be made by knowing the "float" table columns. Since the db->schema's `pdo_type` will be `string`, we need to inspect the raw `type` from [here](https://github.com/bcosca/fatfree-core/blob/c9681ed5b2b5ec34abca1afd9be275d469aa24ec/db/sql.php#L339) within the mapper to trigger the type conversion, unless you have defined the float field someone else in your model (i.e. like Cortex requires). So we could add a preg_match check that looks for floating fields. I have collected some of them in my [schema builder definitions](https://github.com/ikkez/f3-schema-builder/blob/master/lib/db/sql/schema.php#L52-L60) but these are probably not all... I think a good starting point could work like this: `preg_match('/float|decimal|real|numeric|dec|double/i',$type)`

**Comment by bcosca** [comment]
There might be a danger here in automating conversion of float values to string (and vice-versa) because we would never know if the app relates to currency conversion.

**Comment by HwapX** [comment]
Why not look to root of the problem?

At least for me the issue is the automatic casting of float to string, in this case ensure number is formatted correctly, perhaps with something like this `is_float($val) ? rtrim(sprintf("%.53F", $val), '0') : (string)$val` or just dont cast and let PDO driver handle it, any rounding problem that arise will happen on user code.

**Comment by ikkez** [comment]
Yes I think this issue can be resolved for v3.6 and then we're nearly ready to go. I'll have a look at it again ;)

**Comment by eazuka** [comment]
Pls do
